network/phonebook.go

```diff
@@ -35,14 +35,37 @@ type Phonebook interface {
 
 	// UpdateRetryAfter updates the retry-after field for the entries matching the given address
 	UpdateRetryAfter(addr string, retryAfter time.Time)
+
+	// waitAndAddConnectionTime will wait to prevent exceeding connectionsRateLimitingCount.
+	// Then it will register the next connection time.
+	WaitAndAddConnectionTime(addr string,
+		connectionsRateLimitingCount uint, connectionsRateLimitingWindowSeconds uint)
 }
 
 type phonebookData struct {
-	retryAfter time.Time
+	retryAfter            time.Time
+	recentConnectionTimes []time.Time
 }
 
 type phonebookEntries map[string]phonebookData
 
+// PopEarliestTime removes the earliest time from recentConnectionTimes in
+// phonebookData for addr
+// It is expected to be later than ConnectionsRateLimitingWindowSeconds
+func (e *phonebookEntries) popEarliestTime(addr string) {
```

Comment:
I know you didn't write the methods that were here before, but for what it's worth I don't think these methods need to be using pointer receivers. If it were just `(e phonebookEntries)` the map could still be modified from within the method, since maps are a reference type.